unison-cli/src/Unison/Share/Sync.hs

```diff
@@ -326,112 +330,256 @@ dagbfs goal children =
-        -- into those we will download right now, and those we will begin downloading on the next iteration of the loop.
-        let (hashes, nextHashes0) =
-              case NESet.splitAt 50 allHashes of
-                This hs1 -> (hs1, Set.empty)
-                That hs2 -> (hs2, Set.empty) -- impossible, this only happens if we split at 0
-                These hs1 hs2 -> (hs1, NESet.toSet hs2)
-        maybeNextHashes <-
-          fmap NESet.nonEmptySet do
-            doDownload hashes >>= \case
-              Nothing -> pure nextHashes0
-              Just newTempEntities -> do
-                newElaboratedHashes <- elaborate newTempEntities
-                pure (Set.union newElaboratedHashes nextHashes0)
-        let !newDownloadCount = downloadCount + NESet.size hashes
-        case maybeNextHashes of
-          Nothing -> pure newDownloadCount
-          Just nextHashes -> loop newDownloadCount nextHashes
-   in \hashes0 -> do
-        hashes <- elaborate hashes0
-        case NESet.nonEmptySet hashes of
-          Nothing -> pure 0
-          Just hashes -> loop 0 hashes
-  where
-    elaborate :: NESet Hash32 -> IO (Set Share.HashJWT)
-    elaborate hashes =
-      Sqlite.runTransaction conn (elaborateHashes hashes)
-
--- | Download a set of entities from Unison Share. Returns the subset of those entities that we stored in temp storage
--- (`temp_entitiy`) instead of main storage (`object` / `causal`) due to missing dependencies.
-downloadEntities ::
   AuthenticatedHttpClient ->
   BaseUrl ->
-  Sqlite.Connection ->
+  (forall a. (Sqlite.Connection -> IO a) -> IO a) ->
   Share.RepoName ->
-  NESet Share.HashJWT ->
-  IO (Maybe (NESet Hash32))
-downloadEntities httpClient unisonShareUrl conn repoName hashes = do
-  Share.DownloadEntitiesSuccess entities <-
-    httpDownloadEntities
-      httpClient
-      unisonShareUrl
-      Share.DownloadEntitiesRequest {repoName, hashes}
-  fmap NESet.nonEmptySet do
-    Sqlite.runTransaction conn do
-      NEMap.toList entities & foldMapM \(hash, entity) ->
-        upsertEntitySomewhere hash entity <&> \case
-          Q.EntityInMainStorage -> Set.empty
-          Q.EntityInTempStorage -> Set.singleton hash
+  DownloadProgressCallbacks ->
+  NESet Hash32 ->
+  IO ()
+completeTempEntities httpClient unisonShareUrl connect repoName callbacks initialNewTempEntities = do
+  -- The set of hashes we still need to download
+  hashesVar <- newTVarIO Set.empty
+
+  -- The set of hashes that we haven't inserted yet, but will soon, because we've committed to downloading them.
+  uninsertedHashesVar <- newTVarIO Set.empty
+
+  -- The entities payloads (along with the jwts that we used to download them) that we've downloaded
+  entitiesQueue <- newTQueueIO
+
+  -- The sets of new (at the time of inserting, anyway) temp entity rows, which we need to elaborate, then download.
+  newTempEntitiesQueue <- newTQueueIO
+
+  workerCount <- newWorkerCount
+
+  Ki.scoped \scope -> do
```

Comment:
This works really nicely 👍🏼  Good work on this lib

Reply:
Thanks!